### DISCUSSION TITLE
Using username for rodauth features

### DISCUSSION THREAD

**Discussion by doloboyz** [jeremyevans/rodauth#422] [General]
Hello,

I was wondering how I would go about allowing rodauth to use either an email or a username within the accounts table for features like login and forgot password.

**Comment by jeremyevans** [comment]
You can set `login_uses_email? false` to not require logins be emails.  However, the reset password, lockout, verify account, and related features would at least need a custom `email_to` configuration so they know where to send emails for the account.

**Comment by janko** [comment]
You can override `account_from_login` to allow either username or email:

```rb
account_from_login do |login|
  # handle the case when login parameter is a username
  unless login.include?("@")
    login = db[:accounts].where(username: login).get(:email)
  end

  super(login)
end
```

**Reply by doloboyz** [reply]
Would this approach still require custom configuration for other rodauth features as Jeremy described above?

**Reply by jeremyevans** [reply]
No.  From Rodauth's perspective, your login would be your email, and that's the default for `email_to`.  However, if you provide a username when logging in, Rodauth would translate it to the related email.  There would still be some custom work to set the :username field for accounts.
